### ISSUE TITLE
hxparse Might Be Useful for Hank

### ISSUE THREAD

**Issue by zicklag** [NQNStudios/hank#17]
[hxparse](https://github.com/Simn/hxparse) could be a useful tool for writing Hank. If it makes the parser code more elegant, then it could make it easier to add syntactic elements and features.

I'm not sure what it adds, but [mo](https://github.com/skial/mo) is also built on top of hxparse and provides helpers for lexing and parsing.

**Comment by zicklag** [comment]
I updated the [gist](https://gist.github.com/zicklag/c2a6060452759ce13864e43135e856f3) with what I was able to manage for the Ink lexer, but I don't have any more time to spend on it, but it could still be used as a base if somebody wants to try.

**Comment by NQNStudios** [comment]
The amount of language features I plan on adding at this point are small. If it becomes significant enough to want to refactor the current parsing mechanism, then I might consider hxparse, but for now I'm gonna close this issue.

**Comment by NQNStudios** [comment]
@Madpumpkin and I are going to try to clean up the parsing mechanism tomorrow at a local co-working session, so I'm reopening this. We may or may not use `hxparse`, but this issue will be the home of the 3rd iteration of Hank's parsing mechanism.

**Comment by NQNStudios** [comment]
hxparse.Lexer throws a compiler error with Haxe 4.0. To continue experimenting with this I've installed Haxe 3.4.7 and Hank may have to specify 3.4.7 as a required version.

**Comment by zicklag** [comment]
If it throws an error for Haxe 4 for you should probably open an issue. It looks like an [issue](https://github.com/Simn/hxparse/issues/51) with the `final` keyword was fixed for Haxe 4 so it probably worked at one point.

**Comment by NQNStudios** [comment]
I've been working on an hxparse reimplementation, and doing some reading to figure out the best possible approach. Here's an article that made a strong impression: [here](https://eli.thegreenplace.net/2013/07/16/hand-written-lexer-in-javascript-compared-to-the-regex-based-ones)

I'm realizing that I'd like to have more control over performance and control flow of parsing--`hxparse` handles a lot of things but adds an extra confounding layer that will make it harder for others to contribute, and also adds an opaque dependency. I think I'll push what I have to a new branch, and go back to doing things the old handwritten way (but cleaner).

**Comment by zicklag** [comment]
Yes, when I started on the hxparse implementation that I tried, I found that Ink was a particularly difficult language to parse with hxparse because of the Ink language's lack symbols to define syntax. While hxparse seemed to do decently well with programming languages, Ink is almost more about the *lack* of syntax so that it is more natural to type for a normal user, and that makes it particularly difficult to define solid rules to parse it.

A couple examples of smaller pure Haxe parsers that I've seen are the official Haxe [XML parser](https://github.com/HaxeFoundation/haxe/blob/master/std/haxe/xml/Parser.hx) and a [WebIDL parser](https://github.com/ncannasse/webidl/blob/master/webidl/Parser.hx). Not sure if those might be of any help for reference or not. :man_shrugging: Parsers seem kind of difficult.
